caput/fftw.py

```diff
@@ -0,0 +1,341 @@
+
+Examples
+--------
+The core of this module is the :class:`FFT`, which essentially just
+abstracts the :class:`pyfftw:FFTW` in the simplest way.
+
+>>> import numpy as np
+>>> from caput import fftw
+>>>
+>>> shape = (24, 50)
+>>> x = np.random.rand(*shape) + 1j * np.random.rand(*shape)
+>>>
+>>> fftobj = fftw.FFT(x.shape, x.dtype, axes=-1)
+>>>
+>>> X = fftobj.fft(x)
+>>> xi = fftobj.ifft(X)
+>>>
+>>> np.allclose(x, xi)
+True
+
+The direct API can also be used, although it is slower when doing repeated
+transforms for arrays of the same shape and type because a new :class:`FFT`
+has to be created each time.
+
+References
+----------
+.. https://pyfftw.readthedocs.io
+.. http://www.fftw.org
+
+Classes
+=======
+- :py:class:`FFT`
+
+Functions
+=========
+- :py:meth:`fft`
+- :py:meth:`ifft`
+- :py:meth:`fftconvolve`
+- :py:meth:`fftwindow`
+"""
+
+from __future__ import annotations
+
+import pyfftw
+
+from caput import mpiutil
+
+
+class FFT:
+    """Faster FFTs with FFTW."""
+
+    def __init__(
+        self,
+        shape: tuple,
+        dtype: type,
+        axes: None | int | tuple = None,
+        forward: bool = True,
+        backward: bool = True,
+    ):
+        """Create FFTW objects for repeat use.
+
+        This implementation is most efficient when used to repeatedly
+        apply ffts to arrays with the same shape and dtype, because a
+        single, highly optimised pathway can be used with a single
+        initialisation.
+
+        Even for a single use this will typically
+        be faster than the `scipy.fft` or `numpy.fft` implementations,
+        especially when multiple cores can be used.
+
+        Parameters
+        ----------
+        shape
+            The shape of the arrays to initialise for
+        dtype
+            Datatype to create a pathway for. At the moment, only
+            complex -> complex or real -> real are supported. The
+            `pyfftw` implementation of the real -> real backward
+            transform will destroy the input array
+        axes
+            Axes over which to apply the fft. Default is all axes.
+        forward
+            If true, initialise the forward fft. Default is True.
+        backward
+            If true, initialise the backward fft. Default is True.
+        """
+        ncpu = mpiutil.cpu_count()
+        self._nsimd = pyfftw.simd_alignment
+        flags = ("FFTW_MEASURE",)
+
+        if axes is None:
+            axes = tuple(range(len(shape)))
+        elif isinstance(axes, int):
+            axes = (axes,)
+
+        # Store fft params
+        self._params = {
+            "ncpu": ncpu,
+            "simd_alignment": self._nsimd,
+            "shape": shape,
+            "dtype": dtype,
+            "axes": axes,
+            "flags": flags,
+        }
+
+        fftargs = {
+            "input_array": pyfftw.empty_aligned(shape, dtype, n=self._nsimd),
+            "output_array": pyfftw.empty_aligned(shape, dtype, n=self._nsimd),
```

Comment:
This seems to be a weird quirk of `pyfftw`, which is one of the various annoyances which prompted making this module. The input/output buffers are used here in determining the optimal pathway for the transform, but when the object is called (i.e. in `self._fft(input, output)`), `pyfftw` called `pyfftw.FFTW.update_arrays`, which _replaces_ the input and output arrays rather than filling the pre-allocated memory. The test here is also done as a doctest

Reply:
FFTW, in general, is nothing if not weird.

Reply:
I haven't looked into it in more detail than this, but I did wonder why it wasn't used, or even available, with numpy/scipy.

Reply:
I suspect it's considered too difficult of a dependency for numpy/scipy.  The C library is not the easiest thing to install.  The FFTW devs discourage installing pre-built binaries because those can't be tuned for the particular machine they're being installed on.